### PR TITLE
#216: Fix binary transparency bug

### PR DESCRIPTION
**Pull request by fono**
After identifying the location of the bug I managed to make a targeted search to find out someone had already bumped into the same problem, and a fix had been provided as well. I don't want to take their credit, so included a comment in the code with a link to the source of the fix (feel free to correct it if this is not the right way to do this).

Thanks for TomBueng and mkl

### CODE REVIEW

**Comment by andreasrosdal** [comment]
@tombueng @mkl-public Can we please use the code you have written here for OpenPDF under the license of OpenPDF, which is LGPL / MPL?

**Comment by andreasrosdal** [comment]
@fono Thanks for submitting this fix. We need the permission from the authors to use this code with a LGPL / MPL license. Or, we need an independent re-implementation developed in full isolation without any direct copying from iText 5.

File: openpdf/src/main/java/com/lowagie/text/Image.java [hunk 1]
```diff
@@ -697,8 +697,16 @@ public static Image getInstance(java.awt.Image image, java.awt.Color color,
                                 transparency[0] = transparency[1] = (transparentPixel >> 16) & 0xff;
                                 transparency[2] = transparency[3] = (transparentPixel >> 8) & 0xff;
                                 transparency[4] = transparency[5] = transparentPixel & 0xff;
+                                // Transparency fix based on https://github.com/itext/itextpdf/pull/20 by tombueng
```
Comment (andreasrosdal) [inline openpdf/src/main/java/com/lowagie/text/Image.java]: Please remove this comment.

Comment (fono) [reply openpdf/src/main/java/com/lowagie/text/Image.java]: OK, removed it.

File: openpdf/src/main/java/com/lowagie/text/Image.java [hunk 2]
```diff
@@ -697,8 +697,15 @@ public static Image getInstance(java.awt.Image image, java.awt.Color color,
                                 transparency[0] = transparency[1] = (transparentPixel >> 16) & 0xff;
                                 transparency[2] = transparency[3] = (transparentPixel >> 8) & 0xff;
                                 transparency[4] = transparency[5] = transparentPixel & 0xff;
+                                for (int prevPixel = 0; prevPixel < j; prevPixel++) {
```
Comment (andreasrosdal) [inline openpdf/src/main/java/com/lowagie/text/Image.java]: Can you please change the name of this variable to something more descriptive? (just find another name, please)

File: openpdf/src/main/java/com/lowagie/text/Image.java [hunk 3]
```diff
@@ -697,8 +697,15 @@ public static Image getInstance(java.awt.Image image, java.awt.Color color,
...
+                                    if ((pixels[prevPixel] & 0xffffff) == transparentPixel) {
+                                        shades = true;
+                                        break;
+                                    }
+                                }
                             }
-                        } else if ((pixels[j] & 0xffffff) != transparentPixel) {
+                        } else if ((pixels[j] & 0xffffff) != transparentPixel && alpha == 0
```
Comment (andreasrosdal) [inline openpdf/src/main/java/com/lowagie/text/Image.java]: Can you please add some parentheses for clarity here?

File: openpdf/src/main/java/com/lowagie/text/Image.java (hunk 2 above)
Comment (andreasrosdal) [inline openpdf/src/main/java/com/lowagie/text/Image.java]: Can you please add a JavaDoc here, with a good description about what is going on here?

**Comment by tombueng** [comment]
Sure you can use it.







Mit freundlichen Grüßen 


Thomas Büngener







On Fri, Aug 9, 2019 at 4:57 PM +0200, "Andreas Rosdal" <notifications@github.com> wrote:












@tombueng @mkl-public Can we please use the code you have written here for OpenPDF under the license of OpenPDF, which is LGPL / MPL?

**Comment by fono** [comment]
Now that this pull request has been merged and closed, shall I resolve your review comments in a separate pull request or I can commit to the same branch and it'll be added automatically? Sorry, I'm not familiar with the common workflow methods here on github.

**Comment by andreasrosdal** [comment]
@fono It's okay, no more changes needed. Thanks.

**Pull request merged**